pop_tools/regrid.py

```diff
@@ -0,0 +1,436 @@
+
+
+def zonal_average(
+    data,
+    grid_name,
+    dest_grid_method='regular_grid',
+    lat_axis=None,
+    lat_axis_bnds=None,
+    lon_axis=None,
+    lon_axis_bnds=None,
+    dx=0.25,
+    dy=0.25,
+):
+    """
+    Computes zonal average of some pop-dataset using some mask grid name.
+
+    Parameters
+    ----------
+    data: `xarray.Dataset`
+           Input dataset containing
+
+    grid_name: str
+           POP-grid name
+
+    dest_grid_method: str, default = 'regular_grid'
+       Method to use for generating the destination grid. Two main options include 'regular grid',
+       which is regular lat-lon grid, or 'lat_axis' which uses the 'lat_aux_grid' specified in the file
+
+    lat_axis: list, array, optional
+       Latitude axis to use for regridding
+
+    lat_axis_bnds: list, array, optional
+       Latitude axis to use for regridding
+
+    lon_axis: list, array, optional
+       Longitudinal axis to use for regridding
+
+    lon_axis_bnds: list, array, optional
+       Longitudinal axis bounds to use for regridding
+
+    dx = float
+         Longitudinal grid spacing for output grid when using regular_grid
+
+    dy = float
+         Latitudinal grid spacing for output grid when using regular grid
+    """
+
+    # Read in the grid file - set coordinates to TLAT and TLONG
+    data = _add_region_mask(data, grid_name)
+
+    # Add the latitude axis if needed
+    if (
+        (dest_grid_method == 'lat_axis')
+        and ('lat_aux_grid' in data.variables)
+        and (lat_axis is None)
+    ):
+        lat_axis_bnds = data.lat_aux_grid.values
+
+    # If no longitude axis specified, calulate the grid corners and use the longitudinal grid in the southern hemisphere
+    # if lon_axis is None:
+    #    lat_c, lon_c = gen_corner_calc(data)
+    #    lon_axis_bnds = lon_c[0, :]
+    #    lon_axis = data.TLONG.values[0, :]
+
+    # Generate the destination grid
+    dst_grid = gen_dest_grid(
+        dest_grid_method, dx, dy, lat_axis_bnds, lat_axis, lon_axis_bnds, lon_axis
+    )
+    print(data, dst_grid)
+
+    # Convert the mask to a uniform grid, using nearest neighbor interpolation
+    mask_regrid = to_uniform_grid(
+        data[['REGION_MASK']], dst_grid, regrid_method='nearest_s2d'
+    ).REGION_MASK
+    weights_regrid = to_uniform_grid(
+        data[['REGION_MASK']], dst_grid, regrid_method='conservative'
+    ).REGION_MASK
+
+    print(np.unique(mask_regrid))
+
+    # Add a mask to the regridding
+    dst_grid['mask'] = (('y', 'x'), (mask_regrid.where(mask_regrid == 0, 1, 0)))
+
+    # Convert the data to a uniform grid, using default conservative regridding
+    data_regrid = to_uniform_grid(data, dst_grid)
+
+    ds_list = [
+        data_regrid.where(mask_regrid == region)
+        .weighted(weights_regrid.where(mask_regrid == region).fillna(0))
+        .mean('x')
+        for region in tqdm(np.unique(mask_regrid))
+        if region != 0
+    ]
```

Comment:
A few comments here:
1. This function should have one job: compute a zonal average. Why should it bring in these regions.
2. Do you really need the `where` on the weights. Also `.where(..., 0)` will save an extra copy.
3. This could be a lot faster if you made `mask_regrid` a 3D variable: (lat, lon, region) and then passed it in as a weight.

Reply:
Yeah... 
1. From the user API perspective, I thought it would be helpful to be able to specify the grid name and handle the logic within here... where would that be better placed? Or what is the higher level api that we would want the user to specify the region information?
2. I can remove the where on the weights
3. That is true; this might also be a better way of dealing with the fractional areas of the masked grid cells...

Reply:
One way would be to expect users to pass in `ds * ds.region_mask` if `region_mask` was a 3D variable `(region, nlat, nlon)`. Would this be equivalent? 

Does the order of operations here: regrid, then multiply by region mask and mean, follow what's in Keith's FORTRAN program?